winch/codegen/src/isa/aarch64/masm.rs

```diff
@@ -114,8 +141,43 @@ impl Masm for MacroAssembler {
         Ok(())
     }
 
-    fn check_stack(&mut self, _vmctx: Reg) -> Result<()> {
-        // TODO: Implement when we have more complete assembler support.
+    fn check_stack(&mut self, vmctx: Reg) -> Result<()> {
+        let ptr_size: u8 = self.ptr_size.bytes().try_into().unwrap();
+
+        // The PatchableAddToReg construct on aarch64 is not a single
+        // add-immediate instruction, but a 3-instruction sequence that loads an
+        // immediate using 2 mov-immediate instructions into _another_ scratch
+        // register before adding it into the target scratch register.
+        //
+        // In other words, to make this work we use _two_ scratch registers, one
+        // to hold the limit we're calculating and one helper that's just used
+        // to load the immediate.
+        //
+        // Luckily on aarch64 we have 2 available scratch registers, ip0 and
+        // ip1.
+
+        let scratch_stk_limit = regs::scratch();
+        let scratch_tmp = regs::ip1();
+
+        self.load_ptr(
+            self.address_at_reg(vmctx, ptr_size.vmcontext_store_context().into())?,
+            writable!(scratch_stk_limit),
+        )?;
+
+        self.load_ptr(
+            Address::offset(
+                scratch_stk_limit,
+                ptr_size.vmstore_context_stack_limit().into(),
+            ),
+            writable!(scratch_stk_limit),
+        )?;
+
+        self.add_stack_max(scratch_stk_limit, scratch_tmp);
+
+        self.asm
+            .subs_rrr(scratch_stk_limit, regs::sp(), OperandSize::S64);
+        self.asm.trapif(Cond::Gt, TrapCode::STACK_OVERFLOW);
```

Comment:
Here we might want to use `self.masm.trapif` instead. Aarch64 requires the SP to be 16-byte aligned, and even though the SP is not used for memory addressing, if a trap is raised, signal handling code will segfault if SP is not correctly aligned. See https://github.com/bytecodealliance/wasmtime/pull/10146 for more details.

Reply:
Oh, sure (`self` already is a `MacroAssembler` so I assume you mean `self.trapif`).